mesh/mesh_readers.cpp

```diff
@@ -886,215 +897,998 @@ void Mesh::ReadInlineMesh(std::istream &input, int generate_edges)
+                     break;
+
+               } // switch (type_of_element)
+            } // for numElements
+         } // for entityBlock
+
+         ReadGmshBoundaryElements(elements_0D,
+                                  elements_1D,
+                                  elements_2D,
+                                  elements_3D);
+
+         MFEM_CONTRACT_VAR(n_partitions);
+         MFEM_CONTRACT_VAR(elem_domain);
+      } // section '$Elements'
+   }  // we reach the end of the file
+}
+
+
+void Mesh::ReadGmshV41(std::istream &input, int binary)
+{
+
+   string buff;
+   // A map between a serial number of the vertex and its number in the file
+   // (there may be gaps in the numbering, and also Gmsh enumerates vertices
+   // starting from 1, not 0)
+   map<int, int> vertices_map;
+
+   // For each point, curve, surface, and volume entities map entity tag to physical tag (physical tag is element material attr)
+   map<int,int> pointEntities_map;
+   map<int,int> curveEntities_map;
+   map<int,int> surfaceEntities_map;
+   map<int,int> volumeEntities_map;
+
+   const int dim3 = 3;
+
+   // Read the lines of the mesh file. If we face specific keyword, we'll treat
+   // the section.
+   //
+   cerr << "Processing gmsh 4.1 format" << endl;
+   while (input >> buff)
+   {
+      if (buff == "$Entities")
+      {
+         unsigned long numPoints, numCurves, numSurfaces, numVolumes;
+         unsigned long numPhysicals, numBrep;
+         int entityTag, physicalTag, brepTag;
+         double boxMin[3], boxMax[3], point[3];
+         bool needBrep, v41=true;
+
+         getline(input,buff);
+
+         if(binary)
+         {
+            input.read(reinterpret_cast<char*>(&numPoints), sizeof(unsigned long));
+            input.read(reinterpret_cast<char*>(&numCurves), sizeof(unsigned long));
+            input.read(reinterpret_cast<char*>(&numSurfaces), sizeof(unsigned long));
+            input.read(reinterpret_cast<char*>(&numVolumes), sizeof(unsigned long));
+         }
+         else
+         {
+            input >> numPoints >> numCurves >> numSurfaces >> numVolumes;
+         }
+
+         cerr << "entities points, lines, surfaces, volumes: " << numPoints << ":" << numCurves << ":" << numSurfaces << ":" << numVolumes << endl;
+
+         ReadGmshEntityMap(input, pointEntities_map, numPoints, needBrep=false, v41=true, binary);
+         ReadGmshEntityMap(input, curveEntities_map, numCurves, needBrep=true, v41=true, binary);
+         ReadGmshEntityMap(input, surfaceEntities_map, numSurfaces, needBrep=true, v41=true, binary);
+         ReadGmshEntityMap(input, volumeEntities_map, numVolumes, needBrep=true, v41=true, binary);
+
+      }// if Entities
+
+      if (buff == "$Nodes") // reading mesh vertices
+      {
+         unsigned long numEntityBlocks, minNodeTag, maxNodeTag;
+         int serial_number, ver=0;
+         double coord[dim3];
+
+         getline(input,buff);
+
+         if (binary)
```

Comment:
Given that this is happening over and over I think it best to pull stuff like this into a bin_io helper.  This guy already exists:
inline T bin_io::read(std::istream& is);

And we could add this guy:
inline T read(std::istream& is, bool binary)

Which make this nicer.

Reply:
Ahh, sounds good!